### PR TITLE
Feature #112118 feat: ACL Implementation

### PR DESCRIPTION
**Pull request by Pratikshakhandagale**
ACL implementation for :
1. Comment
2. Post Story
3. Reports 
4. Poll
5. Event

### CODE REVIEW

**Comment by Pratikshakhandagale** [comment]
@sushan05 - Please review this PR.

File: easysocial/easysocial/comments.php [hunk 1]
```diff
@@ -62,8 +62,16 @@ public function post()
...
-		// Message should not be empty.
-		if (empty($input))
+		// Determine if this user has the permissions to add comment.
+		$access 	= FD::access();
+		$allowed	= $access->get('comments.add');
+
+		if (!$allowed)
+		{
+			$res->result->message = JText::_('PLG_API_EASYSOCIAL_COMMENT_NOT_ALLOW_MESSAGE');
+			$res->result->status = false;
```
Comment (sushan05) [inline easysocial/easysocial/comments.php]: @Pratikshakhandagale  - Use ApiError::raiseError method with proper status code (403) for unauthorised call.

File: easysocial/easysocial/comments.php [hunk 2]
```diff
@@ -159,13 +167,25 @@ public function getComments()
...
-		if (count($data['data']) < 1)
+		// Determine if this user has the permissions to read comment.
+		$access 	= FD::access();
+		$allowed	= $access->get('comments.read');
+
+		if (!$allowed)
 		{
-			$res->empty_message	=	JText::_('APP_USER_KOMENTO_NO_COMMENTS_FOUND');
+			$res->empty_message	=	JText::_('PLG_API_EASYSOCIAL_READ_COMMENT_NOT_ALLOW_MESSAGE');
+			$res->status		=	false;
```
Comment (sushan05) [inline easysocial/easysocial/comments.php]: @Pratikshakhandagale  - Use ApiError::raiseError method with proper status code (403) for unauthorised call.

File: easysocial/easysocial/share.php [hunk 3]
```diff
@@ -50,6 +50,18 @@ public function get()
...
+	}
+
+	/**
+	 * Method description
+	 *
+	 * @return  mixed
+	 *
+	 * @since 1.0
+	 */
+	public function postStory()
```
Comment (sushan05) [inline easysocial/easysocial/share.php]: @Pratikshakhandagale  - except `GET`,  `POST` and other HTTP request methods, all other methods should be private.

File: easysocial/easysocial/comments.php [hunk 4]
```diff
@@ -52,82 +52,79 @@ public function post()
...
-				$table->comment		=	$input;
-				$table->created_by	=	FD::user()->id;
-				$table->created		=	FD::date()->toSQL();
-				$table->parent		=	$parent;
-				$table->params		=	$params;
-				$table->stream_id	=	$streamid;
-				$state				=	$table->store();
-
-				if ($streamid)
+			ApiError::raiseError(403, JText::_('PLG_API_EASYSOCIAL_EMPTY_COMMENT_NOT_ALLOWED_MESSAGE'));
```
Comment (sushan05) [inline easysocial/easysocial/comments.php]: @Pratikshakhandagale - should be 400

File: easysocial/easysocial/comments.php [hunk 5]
```diff
@@ -159,13 +156,24 @@ public function getComments()
...
 
-		if (count($data['data']) < 1)
+		// Determine if this user has the permissions to read comment.
+		$access 	= FD::access();
+		$allowed	= $access->get('comments.read');
+
+		if (!$allowed)
 		{
-			$res->empty_message	=	JText::_('APP_USER_KOMENTO_NO_COMMENTS_FOUND');
+			$res->empty_message = JText::_('PLG_API_EASYSOCIAL_READ_COMMENT_NOT_ALLOW_MESSAGE');
```
Comment (sushan05) [inline easysocial/easysocial/comments.php]: @Pratikshakhandagale  - Use ApiError::raiseError here

File: easysocial/easysocial/comments.php [hunk 6]
```diff
@@ -179,23 +187,17 @@ public function delete()
...
 		$conversion_id	=	$app->input->get('conversation_id', 0, 'INT');
-		$valid			=	1;
 		$result			=	new stdClass;
 
 		if (!$conversion_id)
 		{
-			$result->status		=	0;
-			$result->message	=	JText::_('PLG_API_EASYSOCIAL_INVALID_CONVERSATION_MESSAGE');
-			$valid				=	0;
+			ApiError::raiseError(400, JText::_('COM_EASYSOCIAL_GROUPS_NO_ACCESS'));
```
Comment (sushan05) [inline easysocial/easysocial/comments.php]: @Pratikshakhandagale  - do correct the message

File: easysocial/easysocial/group.php [hunk 7]
```diff
@@ -83,6 +83,13 @@ public function post()
 		$apiResponse->result	=	new stdClass;
 		$postValues				=	$input->post->getArray();
 
+		// Check EasySocial extension version
+		$component = JComponentHelper::getComponent('com_easysocial');
+		$extension = JTable::getInstance('extension');
+		$extension->load($component->id);
+		$manifest = new Joomla\Registry\Registry($extension->manifest_cache);
```
Comment (sushan05) [inline easysocial/easysocial/group.php]: @Pratikshakhandagale  - 
first import this
jimport( 'joomla.registry.registry' );

then  create object by  - 
new JRegistry($data);

File: easysocial/easysocial/votes.php [hunk 8]
```diff
@@ -65,16 +65,26 @@ public function getPollData()
...
+	private function processVote()
 	{
-		$app = JFactory::getApplication();
-		$pollId = $app->input->get('id', 0, 'int');
-		$itemId  = $app->input->get('itemId', 0, 'int');
-		$action  = $app->input->get('act', '', 'default');
+		$app		=	JFactory::getApplication();
+		$pollId		=	$app->input->get('id', 0, 'int');
+		$itemId		=	$app->input->get('itemId', 0, 'int');
+		$action		=	$app->input->get('act', '', 'default');
```
Comment (sushan05) [inline easysocial/easysocial/votes.php]: @Pratikshakhandagale  - There noting like `default` filter. Use on of these - https://docs.joomla.org/Retrieving_request_data_using_JInput

File: easysocial/easysocial/report.php [hunk 9]
```diff
@@ -55,7 +53,7 @@ public function post()
 	 *
 	 * @since 1.0
 	 */
-	public function create_report()
+	private function create_report()
```
Comment (sushan05) [inline easysocial/easysocial/report.php]: @Pratikshakhandagale  - use Camel case for function name

File: easysocial/easysocial/videos.php [hunk 10]
```diff
@@ -60,7 +60,7 @@ public function post()
 	 *
 	 * @since 1.0
 	 */
-	public function get_videos()
+	private function get_videos()
```
Comment (sushan05) [inline easysocial/easysocial/videos.php]: @Pratikshakhandagale - Use camel case for function name

File: easysocial/easysocial/videos_link.php [hunk 11]
```diff
@@ -41,7 +41,7 @@ public function post()
 	 * 	 
 	 * @return  JSON	 
 	 */
-	public function save_video()
+	private function save_video()
```
Comment (sushan05) [inline easysocial/easysocial/videos_link.php]: @Pratikshakhandagale  - Camel case

File: easysocial/easysocial/groups.php [hunk 12]
```diff
@@ -149,10 +149,16 @@ public function post()
...
 
-			$apiResponse->empty_message = JText::_('COM_API_GROUPS_EMPTY_CATEGORY');
+			$apiResponse->empty_message = JText::_('PLG_API_GROUPS_EMPTY_CATEGORY');
+		}
+
+		$groups	=	$model->getGroups($options);
+
+		if ($limit)
+		{
+			$groups		=	array_slice($groups, $limitstart, $limit);
```
Comment (thite-amol) [inline easysocial/easysocial/groups.php]: @Pratikshakhandagale Remove this slice code. Use group method for this.

File: easysocial/easysocial/comments.php [hunk 13]
```diff
@@ -52,82 +52,79 @@ public function post()
...
-			$res->result->status		=	0;
-			$res->result->message	=	JText::_('PLG_API_EASYSOCIAL_EMPTY_ELEMENT_NOT_ALLOWED_MESSAGE');
-			$valid				=	0;
+			ApiError::raiseError(400, JText::_('PLG_API_EASYSOCIAL_EMPTY_ELEMENT_NOT_ALLOWED_MESSAGE'));
 		}
 
-		// Message should not be empty.
-		if (empty($input))
+		// Determine if this user has the permissions to add comment.
+		$access 	= FD::access();
```
Comment (thite-amol) [inline easysocial/easysocial/comments.php]: @Pratikshakhandagale FD is deprecated. Use ES instead.

File: easysocial/easysocial/group.php [hunk 14]
```diff
@@ -83,6 +81,13 @@ public function post()
 		$apiResponse->result	=	new stdClass;
 		$postValues				=	$input->post->getArray();
 
+		// Check EasySocial extension version
+		$component = JComponentHelper::getComponent('com_easysocial');
+		$extension = JTable::getInstance('extension');
+		$extension->load($component->id);
```
Comment (thite-amol) [inline easysocial/easysocial/group.php]: @Pratikshakhandagale Easysocial has their own method to check the version. can you use that

File: easysocial/easysocial/group.php [hunk 15]
```diff
@@ -146,7 +154,18 @@ public function post()
...
 			$options				=	array();
-			$options['workflow_id']	=	$category->getWorkflow()->id;
+
+			if ( $version >= '2.1.0')
+			{
+				$options['workflow_id']	=	$category->getWorkflow()->id;
+			}
+			else
+			{
+				$stepsModel	=	FD::model('Steps');
```
Comment (thite-amol) [inline easysocial/easysocial/group.php]: @Pratikshakhandagale Use ES also check for other files

**Pull request merged**